File: src/main/java/frc/robot/subsystems/DriveSubsystemNew.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems;

import org.wpilib.commands3.Mechanism;
import org.wpilib.commands3.Scheduler;

import edu.wpi.first.epilogue.Logged;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.kinematics.MecanumDriveWheelPositions;
import edu.wpi.first.math.kinematics.MecanumDriveWheelSpeeds;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.units.measure.Current;
import edu.wpi.first.units.measure.LinearVelocity;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.ExpansionHubMotor;
import edu.wpi.first.wpilibj.I2C.Port;
import frc.robot.Constants.DriveConstants;
import frc.robot.subsystems.GoBildaPinpoint.EncoderDirection;
import frc.robot.subsystems.GoBildaPinpoint.GoBildaOdometryPods;

@Logged
public class DriveSubsystemNew extends Mechanism {
  private final ExpansionHubMotor m_frontLeftMotor = new ExpansionHubMotor(0, DriveConstants.kFrontLeftMotorPort);
  private final ExpansionHubMotor m_frontRightMotor = new ExpansionHubMotor(0, DriveConstants.kFrontRightMotorPort);
  private final ExpansionHubMotor m_rearLeftMotor = new ExpansionHubMotor(0, DriveConstants.kRearLeftMotorPort);
  private final ExpansionHubMotor m_rearRightMotor = new ExpansionHubMotor(0, DriveConstants.kRearRightMotorPort);

  private final GoBildaPinpoint m_pinpoint = new GoBildaPinpoint(Port.kPort1);

  private static void setPids(ExpansionHubMotor motor) {
    motor.setDistancePerCount(DriveConstants.kEncoderDistancePerPulse);

    var pidConstants = motor.getVelocityPidConstants();
    pidConstants.setPID(DriveConstants.kP, 0, 0);
    pidConstants.setFF(DriveConstants.kS, DriveConstants.kV, DriveConstants.kA);

    motor.resetEncoder();
    motor.setEnabled(true);
  }

  /** Creates a new DriveSubsystem. */
  public DriveSubsystemNew() {
    m_pinpoint.resetPosAndIMU();

    m_pinpoint.setEncoderResolution(GoBildaOdometryPods.goBILDA_4_BAR_POD);
    m_pinpoint.setOffsets(DriveConstants.kXOffset, DriveConstants.kYOffset);
    m_pinpoint.setEncoderDirections(EncoderDirection.FORWARD, EncoderDirection.REVERSED);

    m_frontLeftMotor.setReversed(DriveConstants.kFrontLeftEncoderReversed);
    m_rearLeftMotor.setReversed(DriveConstants.kRearLeftEncoderReversed);
    m_frontRightMotor.setReversed(DriveConstants.kFrontRightEncoderReversed);
    m_rearRightMotor.setReversed(DriveConstants.kRearRightEncoderReversed);

    setPids(m_frontLeftMotor);
    setPids(m_frontRightMotor);
    setPids(m_rearLeftMotor);
    setPids(m_rearRightMotor);

    Timer.delay(0.5);

    m_pinpoint.update();

    Scheduler.getDefault().addPeriodic(this::periodic);
  }

  public void periodic() {
    m_pinpoint.update();
  }

  public Current getFrontLeftCurrent() {
    return m_frontLeftMotor.getCurrent();
  }

  public Current getFrontRightCurrent() {
    return m_frontRightMotor.getCurrent();
  }

  public Current getRearLeftCurrent() {
    return m_rearLeftMotor.getCurrent();
  }

  public Current getRearRightCurrent() {
    return m_rearRightMotor.getCurrent();
  }

  public boolean isHubConnected() {
    return m_frontLeftMotor.isHubConnected();
  }

  public Pose2d getPose() {
    return m_pinpoint.getPosition();
  }

  // Reset the odometry to the specified pose
  public void resetOdometry(Pose2d pose) {
    m_pinpoint.setPosition(pose);
  }

  public void setSpeeds(MecanumDriveWheelSpeeds speeds) {

    m_frontLeftMotor.setVelocitySetpoint(speeds.frontLeft);
    m_frontRightMotor.setVelocitySetpoint(speeds.frontRight);
    m_rearLeftMotor.setVelocitySetpoint(speeds.rearLeft);
    m_rearRightMotor.setVelocitySetpoint(speeds.rearRight);
  }

  /**
   * Drives the robot at given x, y and theta speeds. Speeds range from [-1, 1]
   * and the linear
   * speeds have no effect on the angular speed.
   *
   * @param xSpeed        Speed of the robot in the x direction
   *                      (forward/backwards).
   * @param ySpeed        Speed of the robot in the y direction (sideways).
   * @param rot           Angular rate of the robot.
   * @param fieldRelative Whether the provided x and y speeds are relative to the
   *                      field.
   */
  public void drive(LinearVelocity xSpeed, LinearVelocity ySpeed, AngularVelocity rot, boolean fieldRelative) {

    var chassisSpeeds = new ChassisSpeeds(xSpeed, ySpeed, rot);
    if (fieldRelative) {
      chassisSpeeds = chassisSpeeds.toRobotRelative(getHeading());
    }
    chassisSpeeds = chassisSpeeds.discretize(0.02);
    var mecanumStates = DriveConstants.kDriveKinematics.toWheelSpeeds(chassisSpeeds);
    mecanumStates = mecanumStates.desaturate(DriveConstants.kMaxWheelSpeed);
    setSpeeds(mecanumStates);
  }

  public void driveJoysticks(double xSpeed, double ySpeed, double rot, boolean fieldRelative) {
    var xSpeedDelivered = DriveConstants.kMaxSpeed.times(xSpeed);
    var ySpeedDelivered = DriveConstants.kMaxSpeed.times(ySpeed);
    var rotDelivered = DriveConstants.kMaxAngularSpeed.times(rot);

    drive(xSpeedDelivered, ySpeedDelivered, rotDelivered, fieldRelative);
  }

  /**
   * Gets the current wheel speeds.
   *
   * @return the current wheel speeds in a MecanumDriveWheelSpeeds object.
   */
  public MecanumDriveWheelSpeeds getCurrentWheelSpeeds() {
    return new MecanumDriveWheelSpeeds(
        m_frontLeftMotor.getEncoderVelocity(),
        m_frontRightMotor.getEncoderVelocity(),
        m_rearLeftMotor.getEncoderVelocity(),
        m_rearRightMotor.getEncoderVelocity());
  }

  /**
   * Gets the current wheel distance measurements.
   *
   * @return the current wheel distance measurements in a
   *         MecanumDriveWheelPositions object.
   */
  public MecanumDriveWheelPositions getCurrentWheelDistances() {
    return new MecanumDriveWheelPositions(
        m_frontLeftMotor.getEncoderPosition(),
        m_frontRightMotor.getEncoderPosition(),
        m_rearLeftMotor.getEncoderPosition(),
        m_rearRightMotor.getEncoderPosition());
  }

  public AngularVelocity getTurnRate() {
    return m_pinpoint.getHeadingVelocity();
  }

  public Rotation2d getHeading() {
    return m_pinpoint.getHeading();
  }

  // @NotLogged
  // private Voltage m_lastVoltage = Volts.of(0);

  // private final SysIdRoutine m_sysIdRoutine =
  //     new SysIdRoutine(
  //         // Empty config defaults to 1 volt/second ramp rate and 7 volt step voltage.
  //         new SysIdRoutine.Config(),
  //         new SysIdRoutine.Mechanism(
  //             // Tell SysId how to plumb the driving voltage to the motors.
  //             voltage -> {
  //               m_frontLeftMotor.setVoltage(voltage);
  //               m_frontRightMotor.setVoltage(voltage);
  //               m_rearLeftMotor.setVoltage(voltage);
  //               m_rearRightMotor.setVoltage(voltage);
  //             },
  //             // Tell SysId how to record a frame of data for each motor on the mechanism being
  //             // characterized.
  //             log -> {
  //               log.motor("drive-left-front")
  //                   .voltage(m_lastVoltage)
  //                   .linearPosition(Meters.of(m_frontLeftMotor.getEncoderPosition()))
  //                   .linearVelocity(MetersPerSecond.of(m_frontLeftMotor.getEncoderVelocity()));
  //               log.motor("drive-left-rear")
  //                   .voltage(m_lastVoltage)
  //                   .linearPosition(Meters.of(m_rearLeftMotor.getEncoderPosition()))
  //                   .linearVelocity(MetersPerSecond.of(m_rearLeftMotor.getEncoderVelocity()));
  //               log.motor("drive-right-front")
  //                   .voltage(m_lastVoltage)
  //                   .linearPosition(Meters.of(m_frontRightMotor.getEncoderPosition()))
  //                   .linearVelocity(MetersPerSecond.of(m_frontRightMotor.getEncoderVelocity()));
  //               log.motor("drive-right-rear")
  //                   .voltage(m_lastVoltage)
  //                   .linearPosition(Meters.of(m_rearRightMotor.getEncoderPosition()))
  //                   .linearVelocity(MetersPerSecond.of(m_rearRightMotor.getEncoderVelocity()));
  //             },
  //             // Tell SysId to make generated commands require this subsystem, suffix test state in
  //             // WPILog with this subsystem's name ("drive")
  //             this));

  // public Command sysIdQuasistatic(SysIdRoutine.Direction direction) {
  //   return m_sysIdRoutine.quasistatic(direction);
  // }

  // public Command sysIdDynamic(SysIdRoutine.Direction direction) {
  //   return m_sysIdRoutine.dynamic(direction);
  // }
}

File: cpp/robot/subsystems/DriveSubsystemNew.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "DriveSubsystemNew.h"

#include <string>

#include <frc/commands3/Command.h>
#include <frc/commands3/Scheduler.h>
#include <frc/kinematics/ChassisSpeeds.h>
#include <frc/system/Timer.h>
#include <units/time.h>

#include "../Constants.h"

using namespace frc::robot::subsystems;

// ---------------------------------------------------------------------------
// Private helper
// ---------------------------------------------------------------------------

void DriveSubsystemNew::SetPids(
    frc::hardware::expansionhub::ExpansionHubMotor& motor) {
  motor.SetDistancePerCount(DriveConstants::kEncoderDistancePerPulse);

  auto pidConstants = motor.GetVelocityPidConstants();
  pidConstants.SetPID(DriveConstants::kP, 0.0, 0.0);
  pidConstants.SetFF(DriveConstants::kS, DriveConstants::kV,
                     DriveConstants::kA);

  motor.ResetEncoder();
  motor.SetEnabled(true);
}

// ---------------------------------------------------------------------------
// Constructor
// ---------------------------------------------------------------------------

DriveSubsystemNew::DriveSubsystemNew() {
  m_pinpoint.ResetPosAndIMU();
  m_pinpoint.SetEncoderResolution(
      GoBildaPinpoint::GoBildaOdometryPods::goBILDA_4_BAR_POD);
  m_pinpoint.SetOffsets(DriveConstants::kXOffset, DriveConstants::kYOffset);
  m_pinpoint.SetEncoderDirections(GoBildaPinpoint::EncoderDirection::FORWARD,
                                   GoBildaPinpoint::EncoderDirection::REVERSED);

  m_frontLeftMotor.SetReversed(DriveConstants::kFrontLeftEncoderReversed);
  m_rearLeftMotor.SetReversed(DriveConstants::kRearLeftEncoderReversed);
  m_frontRightMotor.SetReversed(DriveConstants::kFrontRightEncoderReversed);
  m_rearRightMotor.SetReversed(DriveConstants::kRearRightEncoderReversed);

  SetPids(m_frontLeftMotor);
  SetPids(m_frontRightMotor);
  SetPids(m_rearLeftMotor);
  SetPids(m_rearRightMotor);

  frc::Timer::Delay(0.5_s);

  m_pinpoint.Update();

  frc::commands3::Scheduler::GetDefault().AddPeriodic(
      [this] { Periodic(); });

  frc::MecanumDriveWheelSpeeds zeroSpeeds{};
  SetDefaultCommand(RunRepeatedly([this, zeroSpeeds] { SetSpeeds(zeroSpeeds); })
                        .WithPriority(frc::commands3::Command::kLowestPriority)
                        .Named("Drive Default"));
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

void DriveSubsystemNew::Periodic() {
  m_pinpoint.Update();
}

units::ampere_t DriveSubsystemNew::GetFrontLeftCurrent() {
  return m_frontLeftMotor.GetCurrent();
}

units::ampere_t DriveSubsystemNew::GetFrontRightCurrent() {
  return m_frontRightMotor.GetCurrent();
}

units::ampere_t DriveSubsystemNew::GetRearLeftCurrent() {
  return m_rearLeftMotor.GetCurrent();
}

units::ampere_t DriveSubsystemNew::GetRearRightCurrent() {
  return m_rearRightMotor.GetCurrent();
}

bool DriveSubsystemNew::IsHubConnected() {
  return m_frontLeftMotor.IsHubConnected();
}

frc::Pose2d DriveSubsystemNew::GetPose() {
  return m_pinpoint.GetPosition();
}

void DriveSubsystemNew::ResetOdometry(const frc::Pose2d& pose) {
  m_pinpoint.SetPosition(pose);
}

void DriveSubsystemNew::SetSpeeds(const frc::MecanumDriveWheelSpeeds& speeds) {
  m_frontLeftMotor.SetVelocitySetpoint(speeds.frontLeft.value());
  m_frontRightMotor.SetVelocitySetpoint(speeds.frontRight.value());
  m_rearLeftMotor.SetVelocitySetpoint(speeds.rearLeft.value());
  m_rearRightMotor.SetVelocitySetpoint(speeds.rearRight.value());
}

void DriveSubsystemNew::Drive(units::meters_per_second_t xSpeed,
                               units::meters_per_second_t ySpeed,
                               units::radians_per_second_t rot,
                               bool fieldRelative) {
  frc::ChassisSpeeds chassisSpeeds{xSpeed, ySpeed, rot};
  if (fieldRelative) {
    chassisSpeeds =
        frc::ChassisSpeeds::FromFieldRelativeSpeeds(chassisSpeeds, GetHeading());
  }
  chassisSpeeds = frc::ChassisSpeeds::Discretize(chassisSpeeds, 0.02_s);
  auto mecanumStates =
      DriveConstants::kDriveKinematics.ToWheelSpeeds(chassisSpeeds);
  mecanumStates.Desaturate(DriveConstants::kMaxWheelSpeed);
  SetSpeeds(mecanumStates);
}

void DriveSubsystemNew::DriveJoysticks(double xSpeed, double ySpeed, double rot,
                                        bool fieldRelative) {
  auto xSpeedDelivered = DriveConstants::kMaxSpeed * xSpeed;
  auto ySpeedDelivered = DriveConstants::kMaxSpeed * ySpeed;
  auto rotDelivered = DriveConstants::kMaxAngularSpeed * rot;

  Drive(xSpeedDelivered, ySpeedDelivered, rotDelivered, fieldRelative);
}

frc::MecanumDriveWheelSpeeds DriveSubsystemNew::GetCurrentWheelSpeeds() {
  return frc::MecanumDriveWheelSpeeds{
      units::meters_per_second_t{m_frontLeftMotor.GetEncoderVelocity()},
      units::meters_per_second_t{m_frontRightMotor.GetEncoderVelocity()},
      units::meters_per_second_t{m_rearLeftMotor.GetEncoderVelocity()},
      units::meters_per_second_t{m_rearRightMotor.GetEncoderVelocity()}};
}

frc::MecanumDriveWheelPositions DriveSubsystemNew::GetCurrentWheelDistances() {
  return frc::MecanumDriveWheelPositions{
      units::meter_t{m_frontLeftMotor.GetEncoderPosition()},
      units::meter_t{m_frontRightMotor.GetEncoderPosition()},
      units::meter_t{m_rearLeftMotor.GetEncoderPosition()},
      units::meter_t{m_rearRightMotor.GetEncoderPosition()}};
}

units::radians_per_second_t DriveSubsystemNew::GetTurnRate() {
  return m_pinpoint.GetHeadingVelocity();
}

frc::Rotation2d DriveSubsystemNew::GetHeading() {
  return m_pinpoint.GetHeading();
}

frc::commands3::Command DriveSubsystemNew::GetJoystickDriveCommand(
    frc::driverstation::Gamepad& gamepad) {
  return RunRepeatedly([this, &gamepad] {
           DriveJoysticks(-gamepad.GetLeftY(), -gamepad.GetLeftX(),
                          -gamepad.GetRightX(), true);
         })
      .WithPriority(frc::commands3::Command::kDefaultPriority)
      .Named("Joystick Drive");
}

frc::commands3::Command DriveSubsystemNew::DriveForwardTime(double time) {
  return Run([this, time](auto& c) {
           DriveJoysticks(0.0, 1.0, 0.0, true);
           c.Wait(units::second_t{time});
           DriveJoysticks(0.0, 0.0, 0.0, true);
         })
      .WithPriority(frc::commands3::Command::kDefaultPriority)
      .Named("Drive Forward " + std::to_string(time));
}
